compiler/shared_ast/typing.ml

```diff
@@ -330,24 +343,26 @@ let ty (_, { uf; _ }) = uf
 (** Infers the most permissive type from an expression *)
 let rec typecheck_expr_bottom_up :
     type a m.
+    leave_unresolved:bool ->
     A.decl_ctx ->
     (a, m A.mark) A.gexpr Env.t ->
     (a, m A.mark) A.gexpr ->
     (a, mark) A.boxed_gexpr =
- fun ctx env e ->
-  typecheck_expr_top_down ctx env
+ fun ~leave_unresolved ctx env e ->
+  typecheck_expr_top_down ~leave_unresolved ctx env
     (UnionFind.make (add_pos e (TAny (Any.fresh ()))))
     e
 
 (** Checks whether the expression can be typed with the provided type *)
 and typecheck_expr_top_down :
     type a m.
+    leave_unresolved:bool ->
     A.decl_ctx ->
     (a, m A.mark) A.gexpr Env.t ->
     unionfind_typ ->
     (a, m A.mark) A.gexpr ->
     (a, mark) A.boxed_gexpr =
- fun ctx env tau e ->
+ fun ~leave_unresolved ctx env tau e ->
```

Comment:
You could have saved some verbosity by adding e.g. `let typecheck_expr_top_down = typecheck_expr_top_down ~leave_unresolved ctx in` here; the verbose version might be clearer though.